libass/ass_parse.c

```diff
@@ -959,8 +959,8 @@ void process_karaoke_effects(ASS_Renderer *render_priv)
                 x_start = 1000000;
                 x_end = -1000000;
                 for (cur2 = s1; cur2 <= e1; ++cur2) {
-                    x_start = FFMIN(x_start, d6_to_int(cur2->bbox.x_min + cur2->pos.x));
-                    x_end = FFMAX(x_end, d6_to_int(cur2->bbox.x_max + cur2->pos.x));
+                    x_start = FFMIN(x_start, d6_to_int(cur2->pos.x));
+                    x_end = FFMAX(x_end, d6_to_int(cur2->pos.x + cur2->advance.x));
```

Comment:
It's due to potential problems with negative `effect_timing`. And the source of such problems is the use of `effect_timing > first_pos_x` as a boolean flag (under `EF_KARAOKE_KO` in `render_text()`).

I don't see how d3286d8a5e3a0186457b8596c2809866fa59d052 can have any effect on kara, as it only moves some parameters between bitmap structures.

Reply:
I don’t follow: what problems are there with negative `effect_timing`?

What I’m saying is, for a while, our karaoke was _completely bonkers_ without this hunk, and I imagine this is the original reason you added it. At current master + this PR, it’s just fine without this hunk. The only difference the hunk makes is that it excludes glyph features that protrude far to the left (e. g. on a cursive “J”) or far to the right (e. g. on a cursive “Q” or “k”) from the computed karaoke block width (and conversely includes blank space within the glyph’s advance width, e. g. around “i”). But this hunk brings us further closer to VSFilter, which also uses advance widths instead of bounding boxes here, so it ends up beneficial but for a reason you didn’t expect.

I don’t know how. But it’s been bisected as the cause of karaoke changes in several other issues on our tracker. It may have no relation to this particular hunk though; that was just a guess.

Reply:
> I don’t follow: what problems are there with negative `effect_timing`?

Maybe not negative, but positive instead (`effect_timing = x_start - d6_to_int(cur2->pos.x)` for \k and \ko at start) which was dependent on outline specifics, so `effect_timing > first_pos_x` condition resulted in unstable results.

I think in future we should get rid of `first_pos_x` and introduce some direct enum flag for that.

Reply:
I think I get what you’re saying: \k and \ko, which are instantaneous unlike \kf, could get triggered at slightly offset times due to glyph shape. OK. But this still has nothing to do with transforms, and the real cause is the fact we set `effect_timing` for them to a real coordinate at all: we should instead set it, conceptually, to minus/plus infinity (which _will_ affect transformed text). I’m planning to do that separately, as part of a batch of VSFilter compatibility fixes for karaoke.

Reply:
Transforms have relation with further `effect_timing` calculations. Basically, I've removed accounting for untransformed bounding box from here and added accounting for transformed bounding box at image generation stage.

Another possible solution could be to subtract `x_start` instead of `d6_to_int(cur2->pos.x)` later in this function.

Reply:
Sorry, but I don’t follow you. As far as I can tell, this hunk has nothing to do with transforms. Nothing at all.

Coordinates are untransformed at this point. Transforms happen later.

The only difference between `cur2->bbox.x_min + cur2->pos.x` and `cur2->pos.x` is that one points to the glyph’s _untransformed_ origin point and the other at the glyph’s _untransformed_ leftmost control point.

Reply:
Sorry if that sounded aggressive; that’s not my intention here. I genuinely fail to see how this is related to transforms, so if you insist that it is, I’d love to hear a more detailed explanation.

Reply:
You _never_ work with `effect_timing` in clean transformed coordinates. This may seem like a bug (and it kinda is), but it’s actually the same with traditional VSFilter, so yay. This entire PR’s outcome is probably closer to VSFilter than you think.

They’re untransformed x coordinates here. Then this is added to the bitmap position later, after transforms. And that’s fine.

Reply:
And if we do later change karaoke to use coordinates after transforms (which would make more sense), that’s fine too, but that’ll be a separate change; I don’t see how this hunk might be related to that.

Reply:
> They’re untransformed x coordinates here. Then this is added to the bitmap position later, after transforms.

That's the point. Instead of untransformed bounding box I use transformed image bounding box. I basically replaced `bbox.x_min` with `bm->left`.

Reply:
You are not. You aren’t using transformed coordinates. You’re still adding untransformed coordinates to image positions (which VSFilter does, too).

Try this:
```
Dialogue: 0,0:00:00.00,0:00:03.00,Default,,0,0,0,,{\frz90}some {\kf100}kara {\kf200}).
Dialogue: 0,0:00:00.00,0:00:03.00,Default,,0,0,0,,{\frz0}some {\kf100}kara {\kf200}).
```
The rotated `).` is still getting filled at the same _horizontal_ speed as the nonrotated one.

Reply:
Yes, I'm not using transformed coordinates for animation span, only for leftmost position.
But I can reword that commit message if there is a better description for such change.

Reply:
You may have moved the origin point slightly, but `effect_timing` (at the point in the pipeline where this hunk stands) is a _difference_ between coordinates, and it’s a difference between pre-transform coordinates both before and after this PR. This is only really visible with \kf, but boy, is it visible with \kf.

This hunk doesn’t actually change anything regarding transforms. You could say you’re more accurately mapping the pre-transform coordinate origin onto the final bitmap origin, but then isn’t it actually _less_ accurate? The resulting bitmap’s position that you’re using, if I’m not mistaken, is the post-transform glyphs’ _bounding box’s leftmost point_ (as per `outline_to_bitmap`) if not the bitmap’s own leftmost point (which can only be placed even more to the left of the glyphs’ leftmost point). But this hunk is _replacing_ use of the glyph’s leftmost point by use of the glyph’s origin. So if the glyph is something like a cursive J with a long tail protruding on the left side… then you’re actually shifting the karaoke line to the left.

Reply:
Sorry, missed your last response.

Frankly, I’m still not sure I grasp how transforms are relevant/improved here; let’s try to take a step back.

You’ve named a real issue that this does (probably) fix: \k and \ko with glyphs that have negative `bbox.x_min`. Transforms aren’t necessary for that to happen. I think there’s a better fix for _that:_ just set `effect_timing` to minus infinity (or zero if that works) or plus infinity for \k and \ko depending on the current time. The exact coordinate is not actually needed for \k and \ko. Moreover, \kf should actually _also_ be set to −∞/0 or to +∞ if the karaoke time has not started or completely elapsed (this is another bug that we have).

I want to make a separate commit for that anyway, so perhaps we can drop this hunk from this commit? And then I’ll reintroduce this hunk in another commit anyway but bill it as aimed at VSFilter compatibility 😄 Do you want me to sketch up a PR with my extra karaoke changes first?

Reply:
> I want to make a separate commit for that anyway, so perhaps we can drop this hunk from this commit?

I think we should drop this PR altogether, as it's designed as quick hack to make karaoke somethat work. If there will be the next karaoke PR, then it's better to do it correctly there.

Reply:
But this PR does the other things right. I think.

Reply:
Although I seem to have an alternative “fix \kf” commit in my deep pile of unfinished things… I’ll try to see whether it’s better or at all relevant.